### PR TITLE
Example of a side-by-side draggable and resortable multiselect field

### PR DESCRIPTION
**Pull request by pburma**
This is an example of how to add a side-by-side multiselect feature. Especially useful for database operations that use arrays. This example uses css to show the selected vs available list of items. It is also possible to use a this code to create single list for the purposes of manually reordering an array. 

image

image

The projects includes sample data to load into the multiselect as shown in the screenshots.

### CODE REVIEW

**Comment by pburma** [comment]
image

File: MultiSelect1/clients/MultiSelect1/controllers/page/Start/onStart.js [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+            $('#setLocations').append('<li class="setList" data-id=' + item.id +'>' + item.name + '</li>');
+        });    
+    }
+
+    client.select("locations", {}, function(results){
+        results.forEach(function(item, index){
+            var obj = {name:item.name, id:item.id};
+            containsObject(obj, setLocations);
+        });
+    });
```
Comment (namirfa) [inline MultiSelect1/clients/MultiSelect1/controllers/page/Start/onStart.js]: Needs to be indented

File: MultiSelect1/clients/MultiSelect1/controllers/page/Start/onStart.js [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+    
+    $( function() {
+        $( "#setLocations, #allLocations" ).sortable({
+            connectWith: ".connectedSortable"
+        }).disableSelection();
+    });
+
+    /* Use parameters when loading from a datatable, remove the select on "selected" block
+    
+    var setLocations = parameters.locations;
```
Comment (namirfa) [inline MultiSelect1/clients/MultiSelect1/controllers/page/Start/onStart.js:15]: Is this supposed to be commented out?

File: MultiSelect1/clients/MultiSelect1/controllers/page/Start/onStart.js [hunk 3]
```diff
@@ -0,0 +1,50 @@
+
```
Comment (namirfa) [inline MultiSelect1/clients/MultiSelect1/controllers/page/Start/onStart.js:1]: Generally weird indentation here, looks like everything is indented to start which probably isn't necessary.

Code inside function defs should be indented, function defs should all be indented the same, etc...

File: MultiSelect1/clients/MultiSelect1/controllers/page/Start/onStart.js [hunk 4]
```diff
@@ -0,0 +1,50 @@
...
+    var setLocations = parameters.locations;
+	
+    */
+    
+    client.select("selected", {}, function(setLocations){
+        loadLocations(setLocations);
+    });
+    
+    function loadLocations(setLocations){
+    if(setLocations){
```
Comment (namirfa) [inline MultiSelect1/clients/MultiSelect1/controllers/page/Start/onStart.js]:
```suggestion
    if (setLocations) {
```

File: MultiSelect1/clients/MultiSelect1/controllers/page/Start/onStart.js [hunk 5]
```diff
@@ -0,0 +1,50 @@
...
+        }).disableSelection();
+    });
+
+    /* Use parameters when loading from a datatable, remove the select on "selected" block
+    
+    var setLocations = parameters.locations;
+	
+    */
+    
+    client.select("selected", {}, function(setLocations){
```
Comment (namirfa) [inline MultiSelect1/clients/MultiSelect1/controllers/page/Start/onStart.js:19]:
```suggestion
    client.select("selected", {}, function(setLocations) {
```

File: MultiSelect1/clients/MultiSelect1/controllers/page/Start/onStart.js [hunk 6]
```diff
@@ -0,0 +1,50 @@
...
+    });
+    
+    function loadLocations(setLocations){
+    if(setLocations){
+        setLocations.forEach(function(item, index){
+            $('#setLocations').append('<li class="setList" data-id=' + item.id +'>' + item.name + '</li>');
+        });    
+    }
+
+    client.select("locations", {}, function(results){
```
Comment (namirfa) [inline MultiSelect1/clients/MultiSelect1/controllers/page/Start/onStart.js]:
```suggestion
    client.select("locations", {}, function(results) {
```

File: MultiSelect1/clients/MultiSelect1/controllers/page/Start/onStart.js [hunk 7]
```diff
@@ -0,0 +1,50 @@
...
+        results.forEach(function(item, index){
```
Comment (namirfa) [inline MultiSelect1/clients/MultiSelect1/controllers/page/Start/onStart.js:31]:
```suggestion
        results.forEach(function(item, index) {
```

File: MultiSelect1/clients/MultiSelect1/controllers/page/Start/widgets/Button3527/onClick.js [hunk 8]
```diff
@@ -0,0 +1,8 @@
+	var list = $('#setLocations li').toArray();
```
Comment (namirfa) [inline MultiSelect1/clients/MultiSelect1/controllers/page/Start/widgets/Button3527/onClick.js:1]: Shouldn't be indented.

File: MultiSelect1/clients/MultiSelect1/controllers/page/Start/widgets/Button3527/onClick.js [hunk 9]
```diff
@@ -0,0 +1,8 @@
...
+    var locations = [];
+    list.forEach(function(item, index){
```
Comment (namirfa) [inline MultiSelect1/clients/MultiSelect1/controllers/page/Start/widgets/Button3527/onClick.js]:
```suggestion
    list.forEach(function(item, index) {
```

**Review by namirfa — CHANGES_REQUESTED** [review]
Pretty much only requesting some code-style changes.

One question:
Is the desired result here that after dragging/sorting the elements in both fields, that the state should be saved? Seems like this just offers the ability to move stuff around, but if I reload the client everything jumps back to it's original state.

Would be helpful to clarify what the intended use-case would be for this.

**Comment by namirfa** [comment]
Also, is there a reason why this is called `MultiSelect1` instead of just `MultiSelect`? Are we expecting to have a `MultiSelect2` and so on?

**Comment by pburma** [comment]
> Also, is there a reason why this is called `MultiSelect1` instead of just `MultiSelect`? Are we expecting to have a `MultiSelect2` and so on?

Yes, there is a couple different variations on the multiselect and will plan on adding others, maybe as components.

**Comment by zhangevan** [comment]
@pburma After importing the project and tried it out. I ran into a bug where the left list is completely empty after moving a few items to the right list:
image

**Review by steve-langley — COMMENTED** [review]
I didn't see the error @zhangevan did, it seemed to work for me. My only comment would be that I think it might be helpful to pad it out with some comments so the user knows what you are doing. In particular it might be good to note that this is based on the "sortable" feature in JQuery-UI (with maybe a link to the doc).

Otherwise a naive user might not have a clue how all this is working, since there doesn't seem to be any code in the example that actually implements the sortability functionality.

File: MultiSelect1/clients/MultiSelect1/controllers/page/Start/onStart.js [hunk 10]
```diff
@@ -0,0 +1,50 @@
...
+    }
+
+    client.select("locations", {}, function(results) {
+        results.forEach(function(item, index){
+            var obj = {name:item.name, id:item.id};
+            containsObject(obj, setLocations);
+        });
+    });
+	}
+        function containsObject(obj, list) {
```
Comment (steve-langley) [inline MultiSelect1/clients/MultiSelect1/controllers/page/Start/onStart.js:37]: Little indenting problem here that makes the code harder to read.

**Comment by zhangevan** [comment]
Don't worry about the bug I reported yesterday. I realized I was using an old version of Chrome (version 75) for other testing purposes.  I don't see the problem on other browsers like Edge and Firefox.

**Review by rblumer4 — APPROVED** [review]
Looks like the only thing holding up this PR is a lack of comments.
But it works, so we can merge it in without that.

**Pull request merged**